tools/json_tools/format/format.cpp

```diff
@@ -0,0 +1,88 @@
+#include "json.h"
+
+#include <cstdlib>
+#include <fstream>
+#include <sstream>
+#include <string>
+
+static void format( JsonIn &jsin, JsonOut &jsout )
+{
+    if( jsin.test_array() ) {
+        jsout.start_array();
+        jsin.start_array();
+        while( !jsin.end_array() ) {
+            format( jsin, jsout );
+        }
+        jsout.end_array();
+    } else if( jsin.test_object() ) {
+        jsout.start_object();
+        jsin.start_object();
+        while( !jsin.end_object() ) {
+            std::string name = jsin.get_member_name();
+            jsout.member( name );
+            format( jsin, jsout );
+        }
+        jsout.end_object();
+    } else if( jsin.test_string() ) {
+        std::string str = jsin.get_string();
+        jsout.write( str );
+    } else if( jsin.test_number() ) {
+        // Have to introspect into the string to distinguish integers from floats.
+        // Otherwise they won't serialize correctly.
+        const int start_pos = jsin.tell();
+        double num = jsin.get_float();
+        const int end_pos = jsin.tell();
+        std::string str_form = jsin.substr( start_pos, end_pos - start_pos );
+        if( str_form.find( '.' ) == std::string::npos ) {
+            jsout.write( static_cast<long>( num ) );
+        } else {
+            jsout.write( num );
+        }
+        jsin.seek( end_pos );
+    } else if( jsin.test_bool() ) {
+        bool tf = jsin.get_bool();
+        jsout.write( tf );
+    } else {
+        // Error?
```

Comment:
It's a little tricky to write something meaningful, but I think I managed something.